Add docker files

haydenbalsys opened the pull request:
Added necessary docker files alongside prettier.

## Description
Added the ability to containerize our client directory to be independently ran with docker.

## Motivation and Context
This simplifies the process of running our client facing application in cohesion with other containers

## How Has This Been Tested?
Manual testing
Jest is not necessary for this specific task of containerizing a directory

## Artifacts (if appropriate):

https://github.com/zdigness/puttdle/assets/74757315/22ec92aa-3477-46e9-9250-6ad975fe60bf

AronPerez commented on client/README.md:
```diff
@@ -28,3 +28,10 @@ export default {
 - Optionally add `plugin:@typescript-eslint/stylistic-type-checked`
 - Install [eslint-plugin-react](https://github.com/jsx-eslint/eslint-plugin-react) and add `plugin:react/recommended` & `plugin:react/jsx-runtime` to the `extends` list
+
+## Running with Docker Compose
+
+1. Make sure you have Docker installed on your machine.
+2. Navigate to the root directory of the project.
+3. Run `docker-compose up` to start the frontend services.
+4. Access the client application at `http://localhost:4200`.
```
It doesn't look like we're using the correct setup for `docker-compose` with our two directories, `client/` and `server/`. Please use the following references to do this:
1. [Github Project File Example](https://github.com/adrianhajdin/project_ai_mern_image_generation/pull/51/files)
2. [Let us take an example application with the following structure:](https://www.docker.com/blog/docker-compose-from-local-to-amazon-ecs/)
3. [Using docker-compose](https://www.baeldung.com/ops/multiple-dockerfiles#docker-compose)

AronPerez replied:
We never updated the docs with our changes here @haydenbalsys

AronPerez commented on client/docker-compose.yml:
```diff
@@ -0,0 +1,10 @@
+      dockerfile: Dockerfile
+    ports:
+      - "4200:4200"
+    command: ["npm", "run", "dev"]
```
Why are we running `npm run dev` twice for the `client/`? We're doing it for the last instruction of the Dockerfile and on `up` for `docker-compose`

AronPerez commented on client/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+EXPOSE 4200
+
+# Define the command to run the client application using the "dev" script
+CMD ["npm", "run", "dev"]
```
Why are we running `npm run dev` twice for the `client/`? We're doing it for the last instruction of the Dockerfile and on `up` for `docker-compose`

AronPerez commented on client/package.json:
```diff
@@ -8,6 +8,7 @@
     "dev": "vite",
     "build": "tsc && vite build",
     "lint": "eslint . --ext ts,tsx --report-unused-disable-directives --max-warnings 0",
```
We should be telling linter to be searching the `./src` directory instead of also `assets/**` and `public/**`

AronPerez commented on client/src/global.d.ts:
```diff
@@ -1,6 +1,9 @@
-    };
-  }
\ No newline at end of file
+  const id: {
+    initialize: (options: { client_id: string; callback: () => void }) => void;
+    renderButton: (
+      element: HTMLElement | null,
+      options: { theme: string; size: string; text: string }
+    ) => void;
+  };
```
This doesn't look like it's being lint-tested against. You'll need to either create a new workflow in `.github/workflows` or find a way to add to the existing one. I would choose the former, as it's easier to manage and organize.

AronPerez commented on client/Dockerfile:
It looks like we're going to need to validate this docker file build in GitHub Actions. Create a new workflow in `.github/workflows` using the standards already there to test this project's docker file builds. Also, be sure to update the names of existing workflows to match the respective areas they're testing.

AronPerez commented on client/package.json:
```diff
@@ -8,6 +8,7 @@
+    "prettier": "prettier --ignore-path .gitignore --write src",
     "preview": "vite preview"
   },
   "dependencies": {
```
Side todo or maybe you can do it, but "@type" dependencies should be in `devDependencies`. See the following: [Item 45: Put TypeScript and @types in devDependencies](https://www.oreilly.com/library/view/effective-typescript/9781492053736/ch06.html) for further explanation

AronPerez commented on client/package.json:
Are we not going to mirror the `tsconfig.json` and `.eslintrc.json` in `server/`? Also if we are, we're missing the following deps


```JSON
    "@typescript-eslint/eslint-plugin": "^6.7.2",
    "@typescript-eslint/parser": "^6.7.2",
    "@stylistic/eslint-plugin": "^1.7.0",
    "eslint-config-prettier": "9.1.0",
    "eslint-plugin-prettier": "5.1.3",
```

Let me know your thoughts

AronPerez reviewed (changes requested):
Good first attempt. Let's work on the following:
1. Enforcing the new `lint` rules with `github actions` for the `client/` directory
2. Updating the `docker-compose` set up for the project
3. Enforcing new `dockerfile build` for `client/` and `docker-compose up` for the project with `github actions`
4. Updating linter rules (optional)
- 1. Adding dependencies
- 2. Copying the lint file from `server/` over to `client/` and deleting the existing one +  unused dependencies

Please add the new artifacts as a comment after you've pushed these changes. Don't update the existing description, as we can use it as an audit trail.

zdigness commented on server/src/controllers/__tests__/map.spec.ts:
```diff
@@ -51,7 +51,9 @@ describe("MapController", () => {
       const fullMap: FullMap | null = await MapController.getMap(new Date())
 
-      expect(database.Map.findOne).toHaveBeenCalledWith({ where: { day: mockMap.day } })
+      expect(database.Map.findOne).toHaveBeenCalledWith({
+        where: { day: expect.anything() },
+      })
```
I think this might be a better solution, can you try this and see if it passes the tests.

```JAVASCRIPT
const today = new Date()
const date = new Date(today.getFullYear(), today.getMonth(), today.getDate())
...
const fullMap: FullMap | null = await MapController.getMap(date)
```

zdigness reviewed (changes requested):
One change recommended:

1. New date solution for map.spec.ts

AronPerez commented:
@haydenbalsys You can run `npm run lint` locally in the `client` directory instead testing with pushes

haydenbalsys commented:
## Description

1. Moved docker-compose.yml file to top level to run both server and client containers from root.
2. Added github actions to verify docker-compose and our client linter
3. Added packages to support eslint in client directory
4. Fixed map test to test date and time, removing time to just compare dates

## Motivation and Context
Previous PR was failing tests and github actions were not being used

## How Has This Been Tested?

1. Jest tests and Github actions have verified changes work
2. Linter ran through `src` in `client` directory and applied fixes

## Artifacts (if appropriate):

https://github.com/zdigness/puttdle/assets/74757315/9d346177-ade7-4d8b-bbaa-085cfb89f0a2

AronPerez commented on .github/workflows/docker-build-and-compose.yml:
```diff
@@ -0,0 +1,49 @@
+    - name: Set up Docker Buildx
+      uses: docker/setup-buildx-action@v2
+
+    - name: Build client Docker image
+      uses: docker/build-push-action@v3
+      with:
+        context: ./client
+        file: ./client/Dockerfile
+        push: false
+        tags: client:latest
+
+    - name: Build server Docker image
+      uses: docker/build-push-action@v3
+      with:
+        context: ./server
+        file: ./server/Dockerfile
+        push: false
+        tags: server:latest
```
Why are we not just using `ubuntu-latest` built in `docker build` command?

AronPerez commented on .github/workflows/docker-build-and-compose.yml:
```diff
@@ -0,0 +1,49 @@
+    - name: Run docker-compose
+      run: |
+        docker-compose up -d
+        docker-compose ps
```
The `server/` `docker-compose` was originally running the tests locally, I wonder if there's a way we can validate the containers running from docker-compose ps. Since the compose can fail and github actions will pass, we need to find a way to make sure we are seeing the containers up we expect

AronPerez commented on .github/workflows/docker-build-and-compose.yml:
Let's separate the `docker-compose` workflow so that each directory has an independent `docker build`, similar to our lint workflows. Is it better for us to manage our Docker workflows in one action?

AronPerez commented on .github/workflows/docker-build-and-compose.yml:
```diff
@@ -0,0 +1,49 @@
+  push:
+    branches:
+      - main
+    paths:
+      - 'client/**'
+      - 'server/**'
+      - 'docker-compose.yml'
```
Let's remove the [path qualifiers; they're overly verbose](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-including-and-excluding-paths). Why do we only care about changes to the `client/` and `server/` dirs and the `docker-compose.yml` file? We were already only concerned about running these if a `push` or `pull_request` is made to the `main` branch; that's sufficient.

AronPerez commented on docker-compose.yml:
```diff
@@ -1,10 +1,18 @@
-      context: .
+      context: ./server
       dockerfile: Dockerfile
     environment:
       NODE_ENV: ${NODE_ENV}
```
Can you validate that if we have `./server/.env` setup, it will add the env vars correctly please? A quick video should be good

haydenbalsys replied:
Not this week. Dont have time. Will send this PR out and wont be able to work on anything till Monday, feel free to do it yourself if that timeline is restricting.

AronPerez commented on client/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+# Bundle app source inside Docker image
+COPY . .
+
+# Make port 8000 available to the world outside this container
```
Should say port `4200`

haydenbalsys commented on .github/workflows/lint-server.yml:
```diff
@@ -13,7 +13,7 @@ jobs:
     defaults:
       run:
-        working-directory: ./server
+        working-directory: ./server/src
```
Thought we wanted to specifically lint the src directory but this is just the workflow so it doesnt matter.

AronPerez replied:
We’re already specifying the ./src directory with the lint script in the client and server directory

AronPerez commented on client/package.json:
```diff
@@ -7,7 +7,9 @@
     "dev": "vite",
     "build": "tsc && vite build",
-    "lint": "eslint . --ext ts,tsx --report-unused-disable-directives --max-warnings 0",
+    "lint": "eslint --ext .ts ./src",
```
Why are we not lining React files anymore (.tsx)?

AronPerez commented on .github/workflows/docker-build-and-compose-client.yml:
This is looking strong 💪 

So, we want two Dockerfile build workflows: one for the server and one for the client. 

We want one Docker Compose workflow: this will spin up both the client and server.

Two new workflows should be created: one for the Dockerfile build with the client directory, and one for the updated Docker Compose we use in the root directory.

AronPerez reviewed (changes requested):
Last push, just github action and package.json stuff
